### PR TITLE
CompositeByteBuf.addComponent always assume reference count ownership

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
The current interface for CompositeByteBuf.addComponent is not clear under what conditions ownership is transferred when addComponent is called. There should be a well defined behavior so that users can ensure that no leaks occur.

Modifications:
- CompositeByteBuf.addComponent should always assume reference count ownership

Result:
Users that call CompositeByteBuf.addComponent do not have to independently check if the buffer's ownership has been transferred and if not independently release the buffer.
Fixes https://github.com/netty/netty/issues/4760

### CODE REVIEW

**Comment by Scottmitch** [comment]
@normanmaurer @trustin - PTAL

**Comment by normanmaurer** [comment]
@trustin WDYT ?

File: buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java [hunk 1]
```diff
@@ -15,6 +15,7 @@
  */
 package io.netty.buffer;
 
+import io.netty.util.ReferenceCountUtil;
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java]: unused import ?

File: buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java [hunk 2]
```diff
@@ -217,24 +246,36 @@ public CompositeByteBuf addComponents(int cIndex, ByteBuf... buffers) {
...
-            int size = components.size();
-            if (cIndex > size) {
-                cIndex = size;
+            return cIndex;
+        } finally {
+            for (; i < buffers.length; ++i) {
+                ByteBuf b = buffers[i];
+                if (b != null) {
+                    try {
+                        b.release();
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java]: ReferenceCountUtil.safeRelease(...) ?

File: buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java [hunk 3]
```diff
@@ -253,21 +298,32 @@ public CompositeByteBuf addComponents(int cIndex, Iterable<ByteBuf> buffers) {
...
+                for (ByteBuf b: buffers) {
+                    list.add(b);
+                }
+                buffers = list;
+            } finally {
+                if (buffers != list) {
+                    for (ByteBuf b: buffers) {
+                        if (b != null) {
+                            try {
+                                b.release();
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java]: see above

File: buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java (hunk 1 above)
Comment (Scottmitch) [inline buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java]: I was just anticipating your next comments :wink:

File: buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java (hunk 2 above)
Comment (Scottmitch) [inline buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java]: I originally had this, but figured we could avoid the instanceof check and casting.

**Comment by normanmaurer** [comment]
@trustin WDYT ?

**Comment by Scottmitch** [comment]
@trustin - I'll pull in tomorrow if there are no objections / concerns.

**Comment by normanmaurer** [comment]
@Scottmitch ship it.

**Comment by Scottmitch** [comment]
Cherry-picked 4.1 (6312c2f) 4.0 (998b8a8)

**Pull request closed without merging**